synapse/util/async_helpers.py

```diff
@@ -173,20 +176,26 @@ def concurrently_execute(
         limit: Maximum number of conccurent executions.
 
     Returns:
-        Deferred[list]: Resolved when all function invocations have finished.
+        Deferred: Resolved when all function invocations have finished.
     """
     it = iter(args)
 
-    async def _concurrently_execute_inner():
+    async def _concurrently_execute_inner(value: T) -> None:
         try:
             while True:
-                await maybe_awaitable(func(next(it)))
+                await maybe_awaitable(func(value))
+                value = next(it)
         except StopIteration:
             pass
 
+    # We use `zip` to handle the case where the number of args is less than the
+    # limit, avoiding needlessly spawning unnecessary background tasks.
     return make_deferred_yieldable(
         defer.gatherResults(
-            [run_in_background(_concurrently_execute_inner) for _ in range(limit)],
+            [
+                run_in_background(_concurrently_execute_inner, value)
+                for value, _ in zip(it, range(limit))
```

Comment:
are you sure `zip` is guaranteed not to drop values when one of the iterators runs out of inputs?

https://docs.python.org/3/library/functions.html#zip says:

> `zip()` should only be used with unequal length inputs when you don’t care about trailing, unmatched values from the longer iterables.

indeed:

```
>>> a = [1,2,3,4]
>>> b = [1,2]
>>> 
>>> a_iter = iter(a)
>>> b_iter = iter(b)
>>> list(zip(a_iter, b_iter))
[(1, 1), (2, 2)]
>>> next(a_iter)
4
```

3 went missing :/

Reply:
Oh ugh. Good spot.

Reply:
Hmm, I wonder if:

> Returns an iterator of tuples, where the i-th tuple contains the i-th element from each of the argument sequences or iterables.

means that its safe if we do `zip(range(n), it)`? We probably just want to do it in a different way...

Reply:
Oh, using `itertools.islice(it, limit)` is probably the right way of doing this, and cleaner...